### PR TITLE
specs: add predeploys

### PR DESCRIPTION
**Pull request by tynes**
**Description**
PR to add specs for the predeploys. These need to be upgraded in a backwards compatible way for bedrock, its worth specifying their interfaces since they will always be in the state.

### CODE REVIEW

**Comment by tynes** [comment]
The following script was used to find the existing predeploys:

```ts
import { providers, utils } from 'ethers'
import * as fs from 'fs'

const outfile = 'predeploys.json'

;(async function () {
  const provider = new providers.StaticJsonRpcProvider('https://mainnet.optimism.io')

  const prefixes = [
    '0x420000000000000000000000000000000000',
    '0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD'
  ]


  // we don't need to search the entire space
  // const MAX_U16 = 0xffff
  const predeploys = []

  for (const prefix of prefixes) {
    for (let i = 0; i <= 100; i++) {
      const buf = Buffer.alloc(2);
      buf.writeUInt16BE(i, 0);
      const addr = utils.hexConcat([prefix, utils.hexZeroPad(buf, 2)])
      predeploys.push(addr)
    }
  }

  const detected = []

  for (const predeploy of predeploys) {
    const code = await provider.getCode(predeploy)
    if (code !== '0x') {
      console.log(`Predeploy detected at ${predeploy}`)
      detected.push({
        address: predeploy,
        code: code,
      })
    }
  }

  fs.writeFileSync(outfile, JSON.stringify(detected, null, 2))
})()

```

Results:

```bash
$ cat predeploys.json | jq -r '.[].address'
0x4200000000000000000000000000000000000000
0x4200000000000000000000000000000000000002
0x4200000000000000000000000000000000000006
0x4200000000000000000000000000000000000007
0x420000000000000000000000000000000000000f
0x4200000000000000000000000000000000000010
0x4200000000000000000000000000000000000011
0x4200000000000000000000000000000000000012
0x4200000000000000000000000000000000000013
0xdeaddeaddeaddeaddeaddeaddeaddeaddead0000

```

File: specs/predeploys.md [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+    function setDecimals(uint256 _decimals) public;
+
+    /**
+     * Computes the L1 portion of the fee
+     * based on the size of the RLP encoded tx
+     * and the current l1BaseFee
+     * @param _data Unsigned RLP encoded tx, 6 elements
+     * @return L1 fee that should be paid for the tx
+     */
+    // slither-disable-next-line external-function
```
Comment (maurelian) [inline specs/predeploys.md:530]: Don't need these annotations here.

File: specs/predeploys.md [hunk 2]
```diff
@@ -0,0 +1,669 @@
...
+The following table includes each of the predeploys. The system version
+indicates when the predeploy was introduced. The possible values are `Legacy`
+or `Bedrock`.
+
+| Name                         | Address                                    | System Version |
+| ---------------------------- | ------------------------------------------ | -------------- |
+| OVM\_L2ToL1MessagePasser     | 0x4200000000000000000000000000000000000000 | Legacy         |
+| OVM\_DeployerWhitelist       | 0x4200000000000000000000000000000000000002 | Legacy         |
+| OVM\_ETH                     | 0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000 | Legacy         |
+| WETH                         | 0x4200000000000000000000000000000000000006 | Legacy         |
```
Comment (maurelian) [inline specs/predeploys.md:46]: Nit: this is referred to as WETH9 everywhere else.

**Review by maurelian — COMMENTED** [review]
Looks good. Left a couple nits worth fixing. I'd also suggest updating the glossary here: 

https://github.com/maurelian/optimistic-specs/blob/e398ba7aaa1872188f371de517df9649330a03cd/specs/glossary.md#L104-L112

with this diff: 

```diff
- Optimism has the following predeploys:
-
-- [L1 Attributes Predeployed Contract][l1-attr-predeploy]
+ All predeploy contracts are specified in the [predeploys specification][./predeploys.md].
```

**Comment by tynes** [comment]
Thank you for the review! Addressed your nits

**Comment by tynes** [comment]
For some reason there is a CircleCI build error

```
# No configuration was found in your project. Please refer to https://circleci.com/docs/2.0/ to get started with your configuration.
```

**Comment by trianglesphere** [comment]
> For some reason there is a CircleCI build error
> 
> ```
> # No configuration was found in your project. Please refer to https://circleci.com/docs/2.0/ to get started with your configuration.
> ```

@mslipper Somethings acting up.

@tynes don't worry about circle CI blocking this PR

File: specs/predeploys.md [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+
+The `Withdrawer` is responsible for handling L2 to L1 interactions.
+
+```solidity
+/**
+ * @title Withdrawer
+ * @notice The Withdrawer contract facilitates sending both ETH value and data from L2 to L1.
+ * It is predeployed in the L2 state at address 0x4200000000000000000000000000000000000016.
+ */
+contract Withdrawer {
```
Comment (trianglesphere) [inline specs/predeploys.md:624]: nit: Can we embed contracts directly? My concern is them getting out of date. Another question is will we want to upgrade the pre-deploys? This is pretty out of scope and doesn't need to be addressed as part of this PR.

Comment (tynes) [reply specs/predeploys.md:624]: The point here is that its all interfaces and that we can transparently change the predeploys in a backwards compatible way. I would much rather link to actual solidity files instead of embedding the source code in markdown, so I could add an `interfaces` directory if preferred

Comment (trianglesphere) [reply specs/predeploys.md:624]: Oh nice. That makes a bunch of sense (I just skimmed an saw solidity, not the fact that it was only interfaces).

**Review by maurelian — APPROVED** [review]
LGTM

File: specs/predeploys.md [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+    function isDeployerAllowed(address _deployer) external view returns (bool);
+}
+```
+
+## OVM\_ETH
+
+The `OVM_ETH` contains the ERC20 represented balances of ETH that has been
+deposited to L2. As part of the bedrock upgrade, the balances will be migrated
+from this contract to the actual Ethereum level accounts to preserve EVM
+equivalence.
```
Comment (norswap) [inline specs/predeploys.md:150]: What happens to the public functions here, will they still work?

File: specs/predeploys.md [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+    function transferFrom(
+        address from,
+        address to,
+        uint256 amount
+    ) external returns (bool);
+}
+```
+
+## WETH9
+
```
Comment (norswap) [inline specs/predeploys.md:177]: Description missing?

File: specs/predeploys.md [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+| L2StandardTokenFactory       | 0x4200000000000000000000000000000000000012 | Legacy         |
+| L1BlockNumber                | 0x4200000000000000000000000000000000000013 | Legacy         |
+| GasPriceOracle               | 0x420000000000000000000000000000000000000F | Legacy         |
+| L1Attributes                 | 0x4200000000000000000000000000000000000015 | Bedrock        |
+| Withdrawer                   | 0x4200000000000000000000000000000000000016 | Bedrock        |
+
+## OVM\_L2ToL1MessagePasser
+
+The `OVM_L2ToL1MessagePasser` is part of the legacy bridge. It is being
+deprecated as part of the bedrock upgrade.
```
[the remark this answers is not in the document]
Comment (norswap) [reply specs/predeploys.md:61]: What does deprecation mean? Will it still be usable / work correctly but use is discourage & it could be removed in the future?

File: specs/predeploys.md [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+        address src,
+        address dst,
+        uint wad
+    ) public returns (bool);
+}
+```
+
+## L2CrossDomainMessenger
+
+The `L2CrossDomainMessenger` is part of the legacy bridge system.
```
Comment (norswap) [inline specs/predeploys.md:208]: Same comment as above: is this still functional? What should be used instead?

**Review by norswap — COMMENTED** [review]
Quick not super in depth review. Biggest comment is we should be clearer about the guarantees we're giving, esp. for the deprecated stuff.

File: specs/predeploys.md [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+- [L1Attributes](#l1attributes)
+- [Withdrawer](#withdrawer)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Predeployed smart contracts exist on Optimism at predetermined addresses in
+the genesis state. They are similar to precompiles but instead run directly
+in the EVM instead of running native code outside of the EVM.
```
[the remark this answers is not in the document]
Comment (tynes) [reply specs/predeploys.md:29]: Once we know the height at which we switch to bedrock, we could add that to this document. Maybe it could go into a file dedicated to backwards compatibility?

**Comment by tynes** [comment]
@norswap great feedback, thank you! Will address

File: specs/predeploys.md [hunk 9]
```diff
@@ -0,0 +1,670 @@
...
+- [L2CrossDomainMessenger](#l2crossdomainmessenger)
+- [L2StandardBridge](#l2standardbridge)
+- [SequencerFeeVault](#sequencerfeevault)
+- [L2StandardTokenFactory](#l2standardtokenfactory)
+- [L1BlockNumber](#l1blocknumber)
+- [OVM\_GasPriceOracle](#ovm%5C_gaspriceoracle)
+- [Reserved System Address 1](#reserved-system-address-1)
+- [Reserved System Address 2](#reserved-system-address-2)
+- [L1Attributes](#l1attributes)
+- [Withdrawer](#withdrawer)
```
Comment (tynes) [inline specs/predeploys.md:21]: The `Withdrawer` no longer exists

**Comment by tynes** [comment]
This PR will need to be rebased and updated based on https://github.com/ethereum-optimism/optimistic-specs/pull/421

File: specs/predeploys.md (hunk 4 above)
Comment (tynes) [reply specs/predeploys.md:129]: I've added a note that this contract's usage should be avoided, some important functions (`transfer`) do not work currently

File: specs/predeploys.md [hunk 10]
```diff
@@ -0,0 +1,670 @@
...
+}
+```
+
+## OVM\_DeployerWhitelist
+
+The `OVM_DeployerWhitelist` had arbitrary contract deployment enabled. In the
+legacy system, this contract was hooked into during `CREATE` and `CREATE2` to
+ensure that the deployer was allowlisted. In the bedrock system, this will
+be removed from the `CREATE` codepath since arbitrary contract deployment
+has already been enabled.
```
[the remark this answers is not in the document]
Comment (tynes) [reply specs/predeploys.md]: I've added a note at the top of the document saying that deprecated predeploys should not be used as well as adding a column to the table that includes whether or not a contract is deprecated

File: specs/predeploys.md [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+## OVM\_L2ToL1MessagePasser
+
+The `OVM_L2ToL1MessagePasser` is part of the legacy bridge. It is being
+deprecated as part of the bedrock upgrade.
+
+```solidity
+/**
+ * @title OVM_L2ToL1MessagePasser
+ * @dev The L2 to L1 Message Passer is a utility contract which facilitate an L1 proof of the
+ * of a message on L2. The L1 Cross Domain Messenger performs this proof in its
```
[the remark this answers is not in the document]
Comment (tynes) [reply specs/predeploys.md]: Once the contracts are in the same repo as the specs, we should add links to the contracts themselves as to not need to maintain these here

**Comment by tynes** [comment]
I've pushed a fix to this PR to address comments. I've removed the majority of the interfaces so that we don't need to maintain the interfaces here as well as in the repo. We can link directly to the contracts once they are all in the same repo

**Comment by tynes** [comment]
Closing in favor of https://github.com/ethereum-optimism/optimism/pull/2568
